platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/isolation/IsolatedActionSerializer.kt

```diff
@@ -0,0 +1,180 @@
+import org.gradle.configurationcache.serialization.ClassEncoder
+import org.gradle.configurationcache.serialization.DefaultReadContext
+import org.gradle.configurationcache.serialization.DefaultWriteContext
+import org.gradle.configurationcache.serialization.IsolateOwner
+import org.gradle.configurationcache.serialization.ReadContext
+import org.gradle.configurationcache.serialization.WriteContext
+import org.gradle.configurationcache.serialization.beans.BeanStateReaderLookup
+import org.gradle.configurationcache.serialization.beans.BeanStateWriterLookup
+import org.gradle.configurationcache.serialization.readNonNull
+import org.gradle.configurationcache.serialization.runReadOperation
+import org.gradle.configurationcache.serialization.runWriteOperation
+import org.gradle.configurationcache.serialization.withIsolate
+import org.gradle.configurationcache.services.IsolatedActionCodecsFactory
+import org.gradle.internal.serialize.kryo.KryoBackedDecoder
+import org.gradle.internal.serialize.kryo.KryoBackedEncoder
+import java.io.ByteArrayOutputStream
+import java.io.OutputStream
+import java.util.IdentityHashMap
+
+
+/**
+ * Serialized state of an object graph containing one or more [IsolatedAction]s.
+ *
+ * @param G type of the root object stored in [graph]
+ */
+internal
+class SerializedIsolatedActionGraph<G>(
+    /**
+     * The serialized graph.
+     */
+    val graph: ByteArray,
+
+    /**
+     * External references that are not serialized directly as part of the [graph].
+     * These might include references to classes, value sources and build services.
+     *
+     * Maps the integer written to the serialized [graph] to the external reference.
+     * See [EnvironmentEncoder] and [EnvironmentDecoder] for details.
+     */
+    val environment: Map<Int, Any>,
+)
+
+
+internal
+class IsolatedActionSerializer(
+    private val owner: IsolateOwner,
+    private val beanStateWriterLookup: BeanStateWriterLookup,
+    private val isolatedActionCodecs: IsolatedActionCodecsFactory
+) {
+    fun <G : Any> serialize(action: G): SerializedIsolatedActionGraph<G> {
+        val outputStream = ByteArrayOutputStream()
+        val environmentEncoder = EnvironmentEncoder()
+        serializeTo(outputStream, environmentEncoder, action)
+        return SerializedIsolatedActionGraph(
+            outputStream.toByteArray(),
+            environmentEncoder.getResultingEnvironment()
+        )
+    }
+
+    private
+    fun serializeTo(
+        outputStream: ByteArrayOutputStream,
+        environmentEncoder: EnvironmentEncoder,
+        action: Any
+    ) {
+        writeContextFor(outputStream, environmentEncoder).useToRun {
+            runWriteOperation {
+                withIsolate(owner) {
+                    write(action)
+                }
+            }
+        }
+    }
+
+    private
+    fun writeContextFor(
+        outputStream: OutputStream,
+        classEncoder: ClassEncoder,
+    ) = DefaultWriteContext(
```

Comment:
🤔  Maybe it's time to split up the `configuration-cache` project into multiple. I remember discussions that we potentially should use CC-serialization for everything, so the serialization infra is a good candidate for its own subproject. And then this serialization logic of isolated actions is technically separate from CC, and runs even in the vintage mode.

Reply:
Yes, I made a comment in that direction about the extraction of the `:serialization` subproject (`:serialization => :serialization-base` and we extract cc serialization as `:serialization` given the ADR recommending it to be the default choice of serialization mechanism moving forward).